feat(macos): add `MACOS_DISABLE_TITLEBAR_DRAG` decoration, normalized and refactored decorations

alberti42 opened the pull request:
## Summary

This PR adds a new macOS window decoration flag and refactors the decoration logic for clarity and maintainability.

### New decoration flag

* **`MACOS_DISABLE_TITLEBAR_DRAG`** – when the title is hidden, do **not** reserve the titlebar area for window dragging. This makes the first line of content fully interactive (e.g. easier text selection in the top row).

### Refactor highlights

* Centralized normalization in `resolve_compatible_decorations()` with explicit invariants.
* Clear separation of concerns:
    * `resolve_compatible_decorations` — normalize & enforce invariants.
    * `from_normalized_decoration_to_mask` — build `NSWindowStyleMask` from normalized flags.
    * `apply_decorations_to_window` — apply mask and UI affordances (button visibility, title transparency) using normalized flags.
* Added defensive `debug_assert!` checks to catch misuse of non-normalized flags.
* Improved in-code comments (notably the “titled-by-default” rationale).

---

## Rationale for `MACOS_DISABLE_TITLEBAR_DRAG`

On macOS, hiding the title typically implies enabling `NSFullSizeContentViewWindowMask` so content extends into the titlebar region while AppKit uses that region for dragging. With `MACOS_DISABLE_TITLEBAR_DRAG`, WezTerm **intentionally omits** `NSFullSizeContentViewWindowMask` so that the top-most content becomes fully interactive (no implicit drag region). This is useful for workflows that prioritize selecting/clicking the very first line.

**Tip for users:** macOS supports a system gesture to drag windows from anywhere with **Control + Command + drag**:

```bash
defaults write -g NSWindowShouldDragOnGesture -bool true
# revert:
defaults delete -g NSWindowShouldDragOnGesture
```

With that gesture, it’s often preferable to disable the reserved drag region and keep the top row fully interactive. The same tip for users is provided in the updated documentation.

---

## Behavior & Invariants

After normalization (`resolve_compatible_decorations`), we guarantee:

* If `INTEGRATED_BUTTONS` is present, `MACOS_DISABLE_TITLEBAR_DRAG` is **absent**.
* If `TITLE` is present, `MACOS_DISABLE_TITLEBAR_DRAG` is **absent**.
* If `integrated_title_button_style != MacOsNative`, `INTEGRATED_BUTTONS` is **removed**.

Mask building (`from_normalized_decoration_to_mask`) then uses these invariants to:

* Keep **`NSTitledWindowMask`** as the basis (for consistent AppKit behavior), and later hide the title text/traffic lights when needed.
* Add **`NSResizableWindowMask`** only if `RESIZE` is set.
* Add **`NSFullSizeContentViewWindowMask`** when **either**
    * `TITLE` is hidden **or**
    * `INTEGRATED_BUTTONS` is set,  
        **and** `MACOS_DISABLE_TITLEBAR_DRAG` is **not** set.
* Preserve the prior special case for `MACOS_FORCE_SQUARE_CORNERS` (if only that flag is set), matching legacy behavior.

No behavior changes on Windows/X11/Wayland.

---

## Code Changes (key pieces)

* `resolve_compatible_decorations(...)` — enforce invariants & compatibility.
* `from_normalized_decoration_to_mask(...)` — build mask with clear, additive logic; includes `debug_assert!` sanity checks.
* `apply_decorations_to_window(...)` — normalize first with `resolve_compatible_decorations(...)`, compute mask once with `from_normalized_decoration_to_mask(...)`, then set:
    * standard window button visibility (only if a titlebar exists),
    * title visibility,
    * transparent titlebar rules.

---

## Documentation

* Add `MACOS_DISABLE_TITLEBAR_DRAG` to the **window\_decorations** docs (nightly).
* Document compatibility: ignored if `TITLE` or `INTEGRATED_BUTTONS` is present.
* Provide tip for users for dragging macOS windows using CTRL+Super key combination.

---

## Backwards Compatibility

* Existing configs are unaffected.
* The new flag is opt-in (nightly). If present alongside incompatible flags, normalization removes it (preserving previous behavior).
* Integrated buttons continue to set full-size content, unchanged.

---

## Testing Plan (manual)

**Configurations to verify on macOS:**

* `TITLE | RESIZE` (default)
    * Title visible, full-size content **off**; standard buttons visible; resizable.
* `RESIZE`
    * Title hidden; full-size content **on**; draggable by title area; resizable.
* `RESIZE | MACOS_DISABLE_TITLEBAR_DRAG`
    * Title hidden; full-size content **off**; **no** reserved drag region; resizable.
* `INTEGRATED_BUTTONS | RESIZE` (native style)
    * Integrated buttons visible (in tab bar); full-size content **on**; resizable.
* `TITLE | MACOS_DISABLE_TITLEBAR_DRAG`
    * The disable-drag flag is **ignored**; title visible; normal drag behavior.
* Non-native integrated button style
    * `INTEGRATED_BUTTONS` is removed; behavior matches `TITLE` or `RESIZE` per config.

---

## Follow-ups (optional)

* Unit tests for normalization (pure function): table-driven cases asserting invariant outputs.
* Consider gating or revisiting `MACOS_FORCE_SQUARE_CORNERS` behavior if still experimental (see https://github.com/wezterm/wezterm/pull/6587#issuecomment-3248533942)

---

**Thanks for reviewing the PR!**

alberti42 commented:
I became aware of some limitations to this PR after committing it. Despite these limitations, this PR still represents a significant improvement.

This PR contains two logical pieces:

* **Refactor (core, desired to land):**
    * Normalized decoration handling (`resolve_compatible_decorations`)
    * Additive mask builder (`from_normalized_decoration_to_mask`)
    * Simplified `apply_decorations_to_window`
    * Clear invariants, debug asserts, and extended comments
* **New macOS flag (optional/experimental):**
    * `MACOS_DISABLE_TITLEBAR_DRAG` with some limitations

## Known limitation with the new flag

When `MACOS_DISABLE_TITLEBAR_DRAG` is active **and** `TITLE` is hidden, we omit `NSFullSizeContentViewWindowMask`. In that configuration, AppKit retains the titlebar frame view; together with a transparent titlebar this can show a faint translucent strip at the top. We need more investigation before settling on the final UX/behavior here.

## Request

* **Please focus review primarily on the refactor.** It improves readability, enforces invariants, and should make future changes safer and easier.
* I’m happy to **drop or disable the `MACOS_DISABLE_TITLEBAR_DRAG` flag** from this PR before merge (or immediately after) while keeping all refactoring changes. That would land the structural improvements without committing to the new behavior yet.

## Path forward

* Option A (preferred): **merge refactor**, spin the new flag into a follow-up PR marked experimental, and continue iteration there.
* Option B: **remove the flag from this PR now**, land refactor only; we can reintroduce the flag later once we’ve ironed out the top-strip behavior across macOS versions.

Thanks!

bew commented:
Didn't read the PR yet, but I would recommend Option B, to have a first PR with only the refactoring, and another PR for the flag (that would be merged _after_, with this PR's branch as its target branch to only show the changes for the flag)

alberti42 commented:
Thanks, Bew! I followed your advice and removed for now the decoration MACOS_DISABLE_TITLEBAR_DRAG. I also renamed the PR to make it consistent with the latest changes. Looking around at other terminal emulators, I realized that Kitty succeeds in making the first line selectable while avoiding the title bar. I would have to look at their code to see what trick they play with macOS. It turned out to be trickier than what I originally expected.

I post below a summary of the latest changes. In my view, the code is now ready to be reviewed and, hopefully, later to land upstream.

alberti42 commented:
# PR Report: Refactor macOS Window Decorations Handling

## Overview

In its latest version (commit e3588bdbbd3b7728defdd498039b96c31e8fd34d), this PR refactors the macOS window decoration logic in WezTerm. The goal is to move away from a case-based enumeration of decoration combinations and instead adopt an **additive normalization approach**. This yields a clearer, more maintainable design and enables some extensions and improvements to the original behavior.

---

## Key Improvements

### 1. **Additive `RESIZE` Semantics**
- **Before:** `INTEGRATED_BUTTONS` implicitly made the window resizable, even without the `RESIZE` flag.
- **Now:** `RESIZE` is the *only* flag that controls resizing.
- **Improvement:** Greater flexibility—integrated buttons can be used independently of resizing.

---

### 2. **Explicit Handling of Integrated Buttons**
- **Before:** `TITLE | INTEGRATED_BUTTONS` kept the title visible; integration logic was implicit.
- **Now:** If `INTEGRATED_BUTTONS` is set, the `TITLE` flag is always removed during normalization.  
- **Improvement:** Consistent integrated layout—integrated buttons imply “hidden title,” independent of the `TITLE` flag.

---

### 3. **Square Corners Dominate (`MACOS_FORCE_SQUARE_CORNERS`)**
- **Before:** Special-cased only two combinations:  
  - `MACOS_FORCE_SQUARE_CORNERS`  
  - `MACOS_FORCE_SQUARE_CORNERS | RESIZE`  
  In other mixes, `NSTitledWindowMask` remained.
- **Now:** When `MACOS_FORCE_SQUARE_CORNERS` is set:  
  - Always removes `NSTitledWindowMask` (truly untitled window).  
  - Always removes `TITLE`, `INTEGRATED_BUTTONS`, and `MACOS_USE_BACKGROUND_COLOR_AS_TITLEBAR_COLOR`.  
- **Improvement:** More predictable and principled—“force” consistently dominates. Avoids leaving behind decorations that are meaningless without a titlebar.

---

### 4. **Safer Interaction with AppKit**
- **Before:** Button/visibility setters were always called, even on untitled windows (no-op but unsafe on nil views).  
- **Now:** These setters are skipped entirely when no titlebar exists.  
- **Improvement:** Cleaner, safer interaction with AppKit; avoids calling methods on non-existent views.

---

## Summary
This refactor:
- Simplifies the decoration-to-mask translation by using an **additive model**.  
- Clarifies invariants via normalization (`resolve_compatible_decorations`).  
- Provides **better modularity and predictability** in handling special cases.  
- Removes legacy quirks (e.g., implicit resize via `INTEGRATED_BUTTONS`).  
- Prepares the codebase for **future extension** with minimal risk.

Net effect: the new approach is **clearer, safer, and more consistent**, while preserving expected user-facing behavior except where explicitly improved.

alberti42 commented:
> Didn't read the PR yet, but I would recommend Option B, to have a first PR with only the refactoring, and another PR for the flag (that would be merged _after_, with this PR's branch as its target branch to only show the changes for the flag). Once you did so, please simplify the PR's description to focus on the refactors and their benefits / design choices / ...

Hi Benoit, coming back to your previous reply:

I’ve reintroduced `MACOS_DISABLE_TITLEBAR_DRAG` in commit 8624eb79e51598ea8dc61aec62a6b39c2bc0d8a7 and updated the docs accordingly.

* The flag now works as intended (when the title is hidden, the titlebar region is no longer reserved for dragging).
* Known limitation: the setting is applied at window creation time. Users need to restart WezTerm—or close and reopen a window—for changes to take effect. This is documented, and given how infrequently this option is toggled, I think it’s an acceptable tradeoff for now. We can revisit a live-update approach later.

Let me know if you’d like any tweaks to naming or wording in the docs.

alberti42 commented:
> Will this fix #7197 ?

Hi Benoit, I created a PR #7199 that fixes the issue #7197, which was completely unrelated to the PR discussed here.
